src/search/SearchMessagesScreen.js

```diff
@@ -130,6 +154,50 @@ class SearchMessagesScreenInner extends PureComponent<Props, State> {
     this.handleQuerySubmit(e);
   };
 
+  fetchOlder = async () => {
+    if (
+      this.props.loading
+      || this.state.foundOldest
+      || this.state.isFetching
+      || this.isFetchingOlder
+    ) {
+      return;
+    }
+
+    invariant(this.state.messages !== null, 'must have atleast 1 message already fetched.');
+    this.isFetchingOlder = true;
+    try {
+      const { query } = this.state;
+      const anchor = this.state.messages[0].id;
+      const { messages: fetchedMessages, foundOldest } = await await this.props.dispatch(
+        fetchMessages({
+          anchor,
+          narrow: SEARCH_NARROW(this.state.query),
+          numBefore: config.messagesPerRequest,
+          numAfter: 0,
+        }),
+      );
+      this.setState(prevState => {
+        if (
+          prevState.query === query
+          && prevState.messages
+          && anchor === prevState.messages[0].id
```

Comment:
Flow will tell you to drop the `?.` from my suggested condition at https://github.com/zulip/zulip-mobile/pull/5052#discussion_r755599428 , but it's needed because the array can be empty. So:
```suggestion
          // FlowIssue: ?. is needed because array can be empty
          // flowlint-next-line unnecessary-optional-chain:off
          && prevState.messages?.[0]?.id === anchor
```

(This is basically a bug in flowlint. It stems from Flow deciding to let `foo[i]` where `foo: Array<T>` return `T` rather than `T | void`, which is an understandable pragmatic choice given that programmers usually check the bounds first before getting the element, but is unsound.)

Reply:
With plain `.id` instead of `?.id`, this would blow up if the query successfully completed and found no results — that's what an empty `state.messages` means.

That case is not easy to trigger, because this comes after checking `prevState.query === query` — so the current query is the same one that this `fetchOlder` was started for — and that query must have previously returned some messages for us to get as far as calling `fetchOlder`. (I think; see first comment of this review, above. But once we expand the invariant as per that comment, then the query definitely must have previously returned some messages for us to get past that check.)

But it's a case that can happen. It just needs the query's results to have changed between when we originally tried it (and got some messages that the user scrolled up past to trigger calling `fetchOlder`), and when we tried it again (after changing the query and changing it back, all while this `fetchOlder`'s request was pending.) Specifically, there must have previously been some results, and then those messages disappeared from the search results — they were deleted, or were moved or edited so they no longer matched, or were in a stream we lost access to, maybe some other possibilities.